File: 문제풀이/온라인 저지/프로그래머스/# 02. Level 2/# 172927. 광물 캐기/kotlin/Solution.kt
class Solution {

    private val table = arrayOf(
        mapOf("diamond" to 1, "iron" to 1, "stone" to 1),
        mapOf("diamond" to 5, "iron" to 1, "stone" to 1),
        mapOf("diamond" to 25, "iron" to 5, "stone" to 1)
    )

    fun solution(picks: IntArray, minerals: Array<String>): Int {
        var answer = Integer.MAX_VALUE
        fun dfs(startStem: Int, startIdx: Int) {
            if (startIdx == minerals.size || (picks[0] == 0 && picks[1] == 0 && picks[2] == 0)) {
                if (answer > startStem) answer = startStem
                return
            }
            var c: Int
            var i: Int
            var stem: Int
            for (j in picks.indices) {
                if (picks[j] > 0) {
                    picks[j] --
                    c = 5
                    i = startIdx
                    stem = startStem
                    while (c-- > 0) {
                        stem += table[j][minerals[i]]!!
                        i ++
                        if (i == minerals.size) break
                    }
                    dfs(stem, i)
                    picks[j] ++
                }
            }
        }
        dfs(0, 0)
        return answer
    }

}

File: 문제풀이/온라인 저지/프로그래머스/# 02. Level 2/# 172927. 광물 캐기/python/solution.py
def solution(picks, minerals):
    # 최소 피로도
    min_cost = 25 * 50 + 1

    costs = [{'diamond': 1, 'iron': 1, 'stone': 1}, {'diamond': 5, 'iron': 1, 'stone': 1},
             {'diamond': 25, 'iron': 5, 'stone': 1}]

    def dfs(picks, idx, cost):
        nonlocal min_cost

        ## 모두 채굴했거나, 모든 곡괭이를 다 썼을 경우 종료
        if idx >= len(minerals) or (picks[0] == 0 and picks[1] == 0 and picks[2] == 0):
            # 비용이 최소 비용보다 적을 경우 갱신
            if cost < min_cost:
                min_cost = cost
            return

        # 가진 곡괭이들을 확인
        for i in range(3):

            # 곡괭이를 쓸 수 있다면 확인
            if picks[i] > 0:
                cur = idx
                next_cost = cost
                count = 5
                while cur < len(minerals) and count > 0:
                    next_cost += costs[i][minerals[cur]]
                    cur += 1
                    count -= 1

                # 그렇게 누적한 비용이 최소 비용보다 적으면 더 탐색할 가치가 있으므로 탐색
                # 최소 비용 이상일 경우, 더 탐색할 가치가 없으므로 종료
                if next_cost < min_cost:
                    dfs([picks[i] - 1 if i == j else picks[j] for j in range(3)], cur, next_cost)

    dfs(picks[:], 0, 0)
    return min_cost
